src/event.rs

```diff
@@ -97,210 +94,218 @@ pub enum Key {
-                                    } else {
-                                        MouseEvent::Press(MouseButton::Left, cx, cy)
-                                    }
-                                }
-                                1 => {
-                                    if cb & 0x40 != 0 {
-                                        MouseEvent::Press(MouseButton::WheelDown, cx, cy)
-                                    } else {
-                                        MouseEvent::Press(MouseButton::Middle, cx, cy)
-                                    }
-                                }
-                                2 => MouseEvent::Press(MouseButton::Right, cx, cy),
-                                3 => MouseEvent::Release(cx, cy),
-                                _ => return error,
-                            })
-                        },
-                        Some(Ok(b'<')) => {
-                            // xterm mouse encoding:
-                            // ESC [ < Cb ; Cx ; Cy ; (M or m)
-                            let mut buf = Vec::new();
-                            let mut c = iter.next().unwrap().unwrap();
-                                while match c {
-                                    b'm' | b'M' => false,
-                                    _ => true,
-                                } {
-                                    buf.push(c);
-                                    c = iter.next().unwrap().unwrap();
-                                }
-                            let str_buf = String::from_utf8(buf).unwrap();
-                            let nums = &mut str_buf.split(';');
+                    parse_csi(iter).ok_or(error)?
+                }
+                Some(Ok(c)) => {
+                    let ch = parse_utf8_char(c, iter);
+                    Event::Key(Key::Alt(try!(ch)))
+                }
+                Some(Err(_)) | None => return Err(error),
+            })
+        }
+        b'\n' | b'\r' => Ok(Event::Key(Key::Char('\n'))),
+        b'\t' => Ok(Event::Key(Key::Char('\t'))),
+        b'\x7F' => Ok(Event::Key(Key::Backspace)),
+        c @ b'\x01'...b'\x1A' => Ok(Event::Key(Key::Ctrl((c as u8 - 0x1 + b'a') as char))),
+        c @ b'\x1C'...b'\x1F' => Ok(Event::Key(Key::Ctrl((c as u8 - 0x1C + b'4') as char))),
+        b'\0' => Ok(Event::Key(Key::Null)),
+        c => {
+            Ok({
+                let ch = parse_utf8_char(c, iter);
+                Event::Key(Key::Char(try!(ch)))
+            })
+        }
+    }
+}
 
-                            let cb = nums.next().unwrap().parse::<u16>().unwrap();
-                            let cx = nums.next().unwrap().parse::<u16>().unwrap();
-                            let cy = nums.next().unwrap().parse::<u16>().unwrap();
+/// Parses a CSI sequence, just after reading ^[
+///
+/// Returns None if an unrecognized sequence is found.
+fn parse_csi<I>(iter: &mut I) -> Option<Event>
```

Comment:
Currently returns the parsed event, or `None`.
We could instead return a `Result` if we wanted to avoid unwrapping so much.